File: c++/2504.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

vector<char> stack;

int sum = 0;
int num = 1;

int main() {
  freopen("example.txt", "r", stdin);

  string s;
  cin >> s;

  for (int i = 0; i < s.length(); i++) {
    if (s[i] == '(') {
      stack.push_back(s[i]);
      num *= 2;
    } else if (s[i] == '[') {
      stack.push_back(s[i]);
      num *= 3;
    } else if (s[i] == ')') {
      if (stack.empty() || stack.back() != '(') {
        cout << 0;
        return 0;
      }
      if (s[i - 1] == '(') sum += num;
      stack.pop_back();
      num /= 2;
    } else if (s[i] == ']') {
      if (stack.empty() || stack.back() != '[') {
        cout << 0;
        return 0;
      }
      if (s[i - 1] == '[') sum += num;
      stack.pop_back();
      num /= 3;
    }
  }

  if (stack.empty())
    cout << sum;
  else
    cout << 0;

  return 0;
}

File: js/2504.js
const fs = require('fs');

const s = fs.readFileSync('example.txt').toString().trim();

const stack = [];

let sum = 0;
let num = 1;

let isRight = true;

for (let i = 0; i < s.length; i++) {
  if (s[i] === '(') {
    stack.push(s[i]);
    num *= 2;
  } else if (s[i] === '[') {
    stack.push(s[i]);
    num *= 3;
  } else if (s[i] === ')') {
    if (stack.length === 0 || stack[stack.length - 1] !== '(') {
      isRight = false;
      break;
    }
    if (s[i - 1] === '(') sum += num;
    stack.pop();
    num /= 2;
  } else if (s[i] === ']') {
    if (stack.length === 0 || stack[stack.length - 1] !== '[') {
      isRight = false;
      break;
    }
    if (s[i - 1] === '[') sum += num;
    stack.pop();
    num /= 3;
  }
}

if (stack.length !== 0) isRight = false;

console.log(isRight ? sum : 0);
